feat(v8): Add parameterized queries

lukekim opened the pull request:
This pull request introduces significant improvements for the upcoming v8 release of the Go SDK for Spice.ai. The main changes include comprehensive updates to the documentation to reflect new features and APIs in v8, the addition of detailed instructions for parameterized queries and health checks, expanded testing and benchmarking documentation, and the introduction of a new test file to ensure robust Arrow type handling. Additionally, the GitHub Actions workflow for installing Spice is streamlined for better cross-platform support.

**Documentation and API Updates:**

* Updated `README.md` to reference v8 throughout, including installation instructions, import paths, and GoDoc links. Added a migration guide for upgrading from v7 to v8, and replaced old dataset and test references with new ones. [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L5-R5) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L14-R20) [[3]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R164-R190) [[4]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R204-R307)
* Added extensive documentation for parameterized queries using ADBC, including supported types, typed parameter constructors, and code examples.
* Documented new health check methods (`IsSpiceHealthy`, `IsSpiceReady`) for verifying Spice instance status before executing queries.
* Expanded testing and benchmarking sections with instructions for running various test suites and benchmarks, and listed available benchmarks.

**Testing Improvements:**

* Added a new test file `arrow_types_test.go` with comprehensive tests for Arrow type inference and value appending, covering all supported types and ensuring correct integration with Apache Arrow.

**CI/CD Workflow Enhancements:**

* Simplified the Spice installation step in `.github/workflows/go.yml` by using the same installation script for both Linux and MacOS, removing the separate Homebrew-based MacOS step.
* Updated the workflow to use `spice run` instead of `spiced` for starting the Spice runtime, reflecting current best practices.

lukekim committed:
feat(v8): Add parameterized queries, health checks, and upgrade to Arrow v18

Major Changes:
- Introduce Sql() and SqlWithParams() methods (Query/QueryWithParams still work)
- Add IsSpiceHealthy() and IsSpiceReady() health check endpoints
- Upgrade from Apache Arrow v17 to v18 and Go 1.23 to 1.24

New Features:
- ADBC support with FlightSQL driver for parameterized queries
- Automatic URI scheme detection (grpc+tls:// for cloud, grpc:// for local)
- Comprehensive type handling (Int32/Int64, String/LargeString, Float32/Float64)
- Lazy ADBC client initialization with automatic Prepare() call

Tests & Benchmarks:
- Add comprehensive test coverage for TPC-H customer and lineitem tables
- Add parameterized query tests with multiple data types
- Add 8 benchmark functions covering query, params, health checks, and init
- Add health check tests for cloud and local runtimes
- Fix row counting logic in tests (count rows not record batches)

Documentation:
- Add UPGRADE_V7_TO_V8.md migration guide
- Add v8.0.0 release notes with examples
- Update README with parameterized query examples and benchmarking guide
- Update cmd/main.go with v8 examples

Bug Fixes:
- Fix ADBC Cloud connection by adding proper URI scheme formatting
- Fix parameter binding by calling stmt.Prepare() before Bind()
- Update copyright year to 2025

phillipleblanc commented on adbc.go:
```diff
@@ -0,0 +1,343 @@
+func inferArrowType(val interface{}) (arrow.DataType, error) {
+	switch val.(type) {
+	case int, int8, int16, int32:
+		return arrow.PrimitiveTypes.Int32, nil
```
there are arrow primitive types for int8, int16 and int32

phillipleblanc replied:
similar for uint*

lukekim replied:
@phillipleblanc this should be fixed now.

phillipleblanc commented on client.go:
```diff
@@ -247,3 +258,72 @@ func (c *SpiceClient) getBackoffPolicy() backoff.BackOff {
+// This endpoint requires authentication via API key and returns "ready" in the body when ready.
+// Returns true if ready, false otherwise.
+func (c *SpiceClient) IsSpiceReady(ctx context.Context) bool {
+	if c.apiKey == "" {
```
for a local spice instance, its a valid configuration for the api key to not be set and it will still work.

lukekim replied:
Fixed.

lukekim committed:
Fix lint errors

lukekim committed:
Expand type support

lukekim committed:
Use Go v1.24

lukekim committed:
Add tag

lukekim committed:
Update go-lang-ci

lukekim committed:
Update workflows

lukekim committed:
Remove explicit install of spiced

lukekim committed:
Remove PREPARE support

lukekim committed:
Merge branch 'trunk' into lukim/adbc

lukekim committed:
Merge remote-tracking branch 'origin/trunk' into lukim/adbc

lukekim committed:
Merge remote-tracking branch 'origin/trunk' into lukim/adbc

lukekim committed:
fix: Address PR feedback and improve test reliability

- Add stmt.Prepare() call before Bind() to fix ADBC parameterized queries
- Add QueryWithParams() backward compatibility alias for v7 migration
- Change tests to skip (not fail) when local Spice is not healthy
- Fix IsSpiceReady() to work with local instances without API key
- Update docs to recommend SqlWithParams() instead of QueryWithParams()

lukekim committed:
fix: Replace deprecated reader.Record() with reader.RecordBatch()

lukekim committed:
fix: Change health check failures from Skip to Fatal

lukekim committed:
chore: Update dependencies to latest versions

- apache/arrow-adbc/go/adbc v1.9.0 (was v1.8.0)
- go.opentelemetry.io/otel/* v1.39.0 (was v1.38.0)
- google.golang.org/grpc v1.77.0 (was v1.75.1)
- Always call Prepare() for ADBC connections

lukekim committed:
Update tests

lukekim committed:
Update

lukekim committed:
Fixes

lukekim committed:
Don't skip cloud

lukekim committed:
Update env
